RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+* **Edge runtime and mobile delivery service changes**
+    * Delivery compatibility: communicating operator version table deployed on device and deliver models appropriately
+    * Runtime: load upgraders at model loading time, to ensure older models always work after updating runtime
+    * Unknown operator versions need to throw an error
+    * The operator version and upgraders are built into the runtime for BC.
+    * Allow for the addition of optional keyword-only arguments without a version bump or FC concern
+    * Since additional operators can be introduced in upgraders, tracing based selective build should also cover upgraders: easier for BC because the new runtimes goes with the upgraders. 
+* **torch.package changes**
+    * Each torch.package package contains a table of operators and corresponding version according to PyTorch build used to package the model
+        * Q: How does the torch.package scenario for mapping old versions to current PyTorch operators work?
+        * A: Operator versioning, by design, can’t cover all torch.package use cases. So this should be out of scope.
+* **New documentation required**
+    * e2e BC-breaking guide
+        * To make a BC-breaking change update the version and write a torchscript adaptor and a mobile adaptor
+    * e2e FC-breaking guide
+        * It’s OK to add new optional keyword-only arguments as long as their default semantic preserve the operator’s current semantics
+
+Note that the proposal does not introduce an explicit version to _all_ PyTorch operators. Instead code changes are only required for updated operators with BC/FC breakage, that cannot be handled by automatic BC/FC methods. For other operators, the implicit version is v0.
+
+As an example, there’s a BC/FC breaking update on operator foo.
+
+Before:
+
+foo(Tensor self, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+After:
+
+foo(Tensor self, **Tensor c**, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+
+In schema, a Tensor argument, c is added. Note that it’s not added as a “tailing default argument”, so that BC/FC cannot be handled automatically.
+
+Accordingly, in the kernel of foo, the implementation is updated based on the new argument c. The pseudo code (it’s in Python format, but can be written in C++ as well) looks like:
+
+
+```
+def foo(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!):
+  # The original kernel implementation
+  ...
+  if not c.empty():
+    a.add_(c)
+```
+
+
+
+## Code changes (minimize the work of developer)
+
+If there is a BC/FC break with schema change in a PR, a lint error can be automatically generated, with instructions below to update the PR. 
+
+
+### Version bump
+
+Update version field in _operator_versions.yaml_.
+
+
+### BC updates
+
+The developer needs to implement a BC “upgrader” in Python. The upgrader code is put in a centralized python file, _operator_upgraders.py_. 
```

Comment:
in torchscript

Reply:
I'm wondering if it's easier for developers to write python upgrader, and it's scripted in the backend? It's fine to manually script it and paste the torchscript here though.

Reply:
That would be easier but I'm not sure exactly how it would handle overload resolution

Reply:
Also I don't think we can torchscript Python that is not itself torchscript?